lib/rspec/mocks/proxy.rb

```diff
@@ -98,10 +98,19 @@ def replay_received_message_on(expectation, &block)
 
       # @private
       def check_for_unexpected_arguments(expectation)
-        @messages_received.each do |(method_name, args, _)|
-          next unless expectation.matches_name_but_not_args(method_name, *args)
+        # Return unless message never received any of the specified args
+        return unless all_arguments_unexpected?(expectation)
+        unexpected_args = @messages_received.map { |_method_name, args, _| args }.uniq.flatten
 
-          raise_unexpected_message_args_error(expectation, *args)
+        raise_unexpected_message_args_error(expectation, unexpected_args)
+      end
+
+      # @private
+      def all_arguments_unexpected?(expectation)
```

Comment:
yes but I think it's equivalent here, in fact we can problem remove the return above it too